lib/cswat.rb

```diff
@@ -1829,135 +1837,143 @@ def shift
+              end
             end
+          else
+            # Regular ole unquoted field.
+            csv << (part.empty? ? nil : part)
           end
-        else
-          # Regular ole unquoted field.
-          csv << (part.empty? ? nil : part)
         end
-      end
 
-      # Replace tacked on @col_sep with @row_sep if we are still in an extended
-      # column.
-      csv[-1][-1] = @row_sep if in_extended_col
-
-      if in_extended_col
-        # if we're at eof?(), a quoted field wasn't closed...
-        if @io.eof?
-          raise MalformedCSVError,
-                "Unclosed quoted field on line #{lineno + 1}."
-        elsif @field_size_limit and csv.last.size >= @field_size_limit
-          raise MalformedCSVError, "Field size exceeded on line #{lineno + 1}."
-        end
-        # otherwise, we need to loop and pull some more data to complete the row
-      else
-        @lineno += 1
+        # Replace tacked on @col_sep with @row_sep if we are still in an extended
+        # column.
+        csv[-1][-1] = @row_sep if in_extended_col
+
+        if in_extended_col
+          # if we're at eof?(), a quoted field wasn't closed...
+          if @io.eof?
+            raise MalformedCSVError,
+              "Unclosed quoted field on line #{lineno + 1}."
+          elsif @field_size_limit and csv.last.size >= @field_size_limit
+            raise MalformedCSVError, "Field size exceeded on line #{lineno + 1}."
+          end
+          # otherwise, we need to loop and pull some more data to complete the row
+        else
+          @lineno += 1
 
-        # save fields unconverted fields, if needed...
-        unconverted = csv.dup if @unconverted_fields
+          # save fields unconverted fields, if needed...
+          unconverted = csv.dup if @unconverted_fields
 
-        # convert fields, if needed...
-        csv = convert_fields(csv) unless @use_headers or @converters.empty?
-        # parse out header rows and handle CSV::Row conversions...
-        csv = parse_headers(csv)  if     @use_headers
+          # convert fields, if needed...
+          csv = convert_fields(csv) unless @use_headers or @converters.empty?
+          # parse out header rows and handle CSV::Row conversions...
+          csv = parse_headers(csv)  if     @use_headers
 
-        # inject unconverted fields and accessor, if requested...
-        if @unconverted_fields and not csv.respond_to? :unconverted_fields
-          add_unconverted_fields(csv, unconverted)
-        end
+          # inject unconverted fields and accessor, if requested...
+          if @unconverted_fields and not csv.respond_to? :unconverted_fields
+            add_unconverted_fields(csv, unconverted)
+          end
 
-        # return the results
-        break csv
+          # return the results
+          break csv
+        end
+      rescue MalformedCSVError => e
+        if @graceful_errors
+          return e
```

Comment:
Can be the same thing but there shouldn't be any difference in this case since we aren't doing anything else with the result of loop expression and just returning it. Am I missing something?